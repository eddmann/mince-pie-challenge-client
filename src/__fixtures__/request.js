export default function createRequestStack(responses = []) {
  const push = ({ path, method, body, response }) => responses.push({ path, method, body, response });

  const api = (path, method, body = undefined) => {
    const actual = { path, method, body };

    if (responses.length === 0) {
      return fail(`Expected: Nil, Actual: ${JSON.stringify(actual)}`);
    }

    const { response, ...expected } = responses.shift();

    expect(expected).toEqual(actual);

    return Promise.resolve(response);
  };

  const upload = (url, file) => api(url, 'PUT', file);

  return { push, api, upload };
};
