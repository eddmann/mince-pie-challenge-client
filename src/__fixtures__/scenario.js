export default function createScenario(fixtures = () => ({})) {
  return (label, run) => {
    const steps = [];

    run({
      given: (label, step) => steps.push(step),
      when: (label, step) => steps.push(step),
      then: (label, step) => steps.push(step),
      fixtures: fixtures(),
    });

    it(label, () =>
      steps.reduce((state, step) => state
        .then(wrapper => step(wrapper) || wrapper)
        .then(wrapper => new Promise(res => process.nextTick(() => { wrapper && wrapper.update(); res(wrapper); }))),
      Promise.resolve(undefined)));
  };
};
