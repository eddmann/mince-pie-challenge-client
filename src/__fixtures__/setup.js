import React from 'react';
import Enzyme, { mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import createMemoryHistory from 'history/createBrowserHistory';
import createScenario from './scenario';
import createRequestStack from './request';
import app from '../app';

Enzyme.configure({ adapter: new Adapter() });

export const scenario = createScenario(() => {
  const history = createMemoryHistory();
  const request = createRequestStack();
  const auth = { getToken: () => Promise.resolve() };

  return {
    request,
    auth,
    renderApp: (path = '/') => {
      history.replace(path);
      return app({
        history,
        services: { request, auth },
        links: { add: 'add.url', list: 'list.url', view: 'view.url/{id}' },
      }).then(mount);
    },
  };
});
