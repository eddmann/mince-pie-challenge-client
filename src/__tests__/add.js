import { scenario } from '../__fixtures__/setup';

scenario('adds a mince pie', ({ given, when, then, fixtures: { request, auth, renderApp } }) => {
  const ID = '1234';
  const NAME = 'Super Awesome Mince Pie';
  const PHOTO = { name: 'pie.jpg', type: 'image/jpeg', extension: 'jpg' };
  const REQUEST_URL = 'request.url';
  const UPLOAD_URL = 'upload.url';

  given('a logged in user, on the add page', () => {
    auth.getToken = () => Promise.resolve('SAMPLE_TOKEN');

    request.push({
      path: 'add.url',
      method: 'POST',
      body: { name: NAME },
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          photo: { href: REQUEST_URL },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });

    request.push({
      path: REQUEST_URL,
      method: 'PUT',
      body: { fileExtension: PHOTO.extension, contentType: PHOTO.type },
      response: {
        _links: {
          self: { href: REQUEST_URL },
        },
        url: UPLOAD_URL,
      },
    });

    request.push({
      path: UPLOAD_URL,
      method: 'PUT',
      body: PHOTO,
    });

    request.push({
      path: `view.url/${ID}`,
      method: 'GET',
      response: {
        _links: {
          self: { href: `/pies/${ID}` },
          rate: { href: `/pies/${ID}/rate` },
          remove: { href: `/pies/${ID}` },
        },
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        photo: 'photo.url',
        thumbnail: 'thumbnail.url',
        addedAt: new Date('2018-01-01').toISOString(),
      },
    });

    return renderApp('/add');
  });

  when('add pie with name', wrapper => {
    wrapper.find('input#name').simulate('change', { target: { id: 'name', value: NAME }});
    wrapper.find('Button').simulate('click');
  });

  when('upload pie photo', wrapper => {
    wrapper.find('input').simulate('change', { target: { files: [ PHOTO ] } });
    wrapper.find('Button').simulate('click');
  });

  then('redirected to new pies detail view', wrapper => {
    expect(wrapper.find('CardContent')).toMatchSnapshot();
  });
});
