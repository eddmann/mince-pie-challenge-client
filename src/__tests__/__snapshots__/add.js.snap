// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds a mince pie 1`] = `
Array [
  <CardContent>
    <div
        className="content"
    >
        <CardHeader
            content="Super Awesome Mince Pie"
        >
            <div
                className="header"
            >
                Super Awesome Mince Pie
            </div>
        </CardHeader>
        <CardDescription>
            <div
                className="description"
            >
                Rated 
                0
                /5 from 
                0
                 reviews
            </div>
        </CardDescription>
    </div>
</CardContent>,
  <CardContent
    extra={true}
>
    <div
        className="extra content"
    >
        <Rating
            clearable="auto"
            disabled={false}
            maxRating={5}
            onRate={[Function]}
            size="huge"
        >
            <div
                className="ui huge rating"
                onMouseLeave={[Function]}
                role="radiogroup"
            >
                <RatingIcon
                    active={false}
                    aria-checked={false}
                    aria-posinset={1}
                    aria-setsize={5}
                    as="i"
                    index={0}
                    key="0"
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    selected={false}
                >
                    <i
                        aria-checked={false}
                        aria-posinset={1}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                    />
                </RatingIcon>
                <RatingIcon
                    active={false}
                    aria-checked={false}
                    aria-posinset={2}
                    aria-setsize={5}
                    as="i"
                    index={1}
                    key="1"
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    selected={false}
                >
                    <i
                        aria-checked={false}
                        aria-posinset={2}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                    />
                </RatingIcon>
                <RatingIcon
                    active={false}
                    aria-checked={false}
                    aria-posinset={3}
                    aria-setsize={5}
                    as="i"
                    index={2}
                    key="2"
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    selected={false}
                >
                    <i
                        aria-checked={false}
                        aria-posinset={3}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                    />
                </RatingIcon>
                <RatingIcon
                    active={false}
                    aria-checked={false}
                    aria-posinset={4}
                    aria-setsize={5}
                    as="i"
                    index={3}
                    key="3"
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    selected={false}
                >
                    <i
                        aria-checked={false}
                        aria-posinset={4}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                    />
                </RatingIcon>
                <RatingIcon
                    active={false}
                    aria-checked={false}
                    aria-posinset={5}
                    aria-setsize={5}
                    as="i"
                    index={4}
                    key="4"
                    onClick={[Function]}
                    onMouseEnter={[Function]}
                    selected={false}
                >
                    <i
                        aria-checked={false}
                        aria-posinset={5}
                        aria-setsize={5}
                        className="icon"
                        onClick={[Function]}
                        onKeyUp={[Function]}
                        onMouseEnter={[Function]}
                        role="radio"
                        tabIndex={0}
                    />
                </RatingIcon>
            </div>
        </Rating>
        <Button
            as="button"
            floated="right"
            loading={false}
            negative={true}
            onClick={[Function]}
        >
            <button
                className="ui negative right floated button"
                onClick={[Function]}
                role="button"
            >
                Remove
            </button>
        </Button>
    </div>
</CardContent>,
]
`;
